File: scripts/agent-runtime/active-run.test.mjs
import assert from "node:assert/strict";
import { existsSync, mkdirSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import test from "node:test";
import {
	ACTIVE_RUN_ENV_VAR,
	DEFAULT_RUNS_ROOT,
	resolveRunDir,
} from "./active-run.mjs";
import { RUN_MANIFEST_NAME } from "./workspace-start.mjs";

function withTempBase(fn) {
	const base = mkdtempSync(join(tmpdir(), "cautilus-active-run-"));
	try {
		fn(base);
	} finally {
		rmSync(base, { recursive: true, force: true });
	}
}

test("explicit outputDir wins and is created if missing", () => {
	withTempBase((base) => {
		const target = join(base, "explicit-out");
		const result = resolveRunDir({
			outputDir: target,
			env: { [ACTIVE_RUN_ENV_VAR]: join(base, "ignored") },
			now: new Date("2026-04-11T10:00:00.000Z"),
		});
		assert.equal(result.source, "explicit");
		assert.equal(result.created, false);
		assert.equal(result.runDir, resolve(target));
		assert.equal(existsSync(result.runDir), true);
	});
});

test("explicit outputDir is resolved against an explicit cwd", () => {
	withTempBase((base) => {
		const result = resolveRunDir({
			outputDir: "relative-out",
			cwd: base,
			env: {},
			now: new Date("2026-04-11T10:00:00.000Z"),
		});
		assert.equal(result.runDir, resolve(base, "relative-out"));
		assert.equal(existsSync(result.runDir), true);
	});
});

test("env var is used when no explicit outputDir is provided", () => {
	withTempBase((base) => {
		const active = join(base, "active-from-env");
		mkdirSync(active, { recursive: true });
		const result = resolveRunDir({
			env: { [ACTIVE_RUN_ENV_VAR]: active },
			now: new Date("2026-04-11T10:00:00.000Z"),
		});
		assert.equal(result.source, "active");
		assert.equal(result.created, false);
		assert.equal(result.runDir, resolve(active));
	});
});

test("env var pointing at a missing directory fails loudly", () => {
	withTempBase((base) => {
		assert.throws(
			() =>
				resolveRunDir({
					env: { [ACTIVE_RUN_ENV_VAR]: join(base, "missing") },
					now: new Date("2026-04-11T10:00:00.000Z"),
				}),
			new RegExp(`${ACTIVE_RUN_ENV_VAR} does not exist`),
		);
	});
});

test("env var pointing at a regular file fails loudly", () => {
	withTempBase((base) => {
		const filePath = join(base, "not-a-dir");
		writeFileSync(filePath, "regular file", "utf-8");
		assert.throws(
			() =>
				resolveRunDir({
					env: { [ACTIVE_RUN_ENV_VAR]: filePath },
					now: new Date("2026-04-11T10:00:00.000Z"),
				}),
			new RegExp(`${ACTIVE_RUN_ENV_VAR} must be a directory`),
		);
	});
});

test("auto-materialize creates a fresh runDir under an explicit root", () => {
	withTempBase((base) => {
		const root = join(base, "custom-root");
		const result = resolveRunDir({
			root,
			label: "Mode Held-Out",
			env: {},
			now: new Date("2026-04-11T10:32:15.123Z"),
		});
		assert.equal(result.source, "auto");
		assert.equal(result.created, true);
		assert.equal(result.label, "mode-held-out");
		assert.equal(result.root, resolve(root));
		assert.equal(result.runDir, resolve(root, "20260411T103215123Z-mode-held-out"));
		assert.equal(existsSync(result.runDir), true);
		assert.equal(existsSync(join(result.runDir, RUN_MANIFEST_NAME)), true);
	});
});

test("auto-materialize creates the explicit root recursively when missing", () => {
	withTempBase((base) => {
		const root = join(base, "nested", "missing", "root");
		const result = resolveRunDir({
			root,
			label: "fresh",
			env: {},
			now: new Date("2026-04-11T10:32:15.000Z"),
		});
		assert.equal(result.source, "auto");
		assert.equal(existsSync(result.root), true);
		assert.equal(existsSync(result.runDir), true);
	});
});

test("auto-materialize falls back to the default root relative to cwd", () => {
	withTempBase((base) => {
		const result = resolveRunDir({
			cwd: base,
			label: "fallback",
			env: {},
			now: new Date("2026-04-11T10:32:15.000Z"),
		});
		assert.equal(result.source, "auto");
		assert.equal(result.root, resolve(base, DEFAULT_RUNS_ROOT));
		assert.equal(result.runDir, resolve(base, DEFAULT_RUNS_ROOT, "20260411T103215000Z-fallback"));
		assert.equal(existsSync(result.runDir), true);
	});
});

test("explicit outputDir overrides an active env var", () => {
	withTempBase((base) => {
		const active = join(base, "active");
		mkdirSync(active, { recursive: true });
		const explicit = join(base, "explicit");
		const result = resolveRunDir({
			outputDir: explicit,
			env: { [ACTIVE_RUN_ENV_VAR]: active },
			now: new Date("2026-04-11T10:32:15.000Z"),
		});
		assert.equal(result.source, "explicit");
		assert.equal(result.runDir, resolve(explicit));
	});
});

test("env var beats auto-materialize fallback", () => {
	withTempBase((base) => {
		const active = join(base, "active");
		mkdirSync(active, { recursive: true });
		const result = resolveRunDir({
			cwd: base,
			env: { [ACTIVE_RUN_ENV_VAR]: active },
			now: new Date("2026-04-11T10:32:15.000Z"),
		});
		assert.equal(result.source, "active");
		assert.equal(result.runDir, resolve(active));
		// auto-materialize default root must NOT have been created
		assert.equal(existsSync(resolve(base, DEFAULT_RUNS_ROOT)), false);
	});
});

test("auto-materialize uses the default label when none is provided", () => {
	withTempBase((base) => {
		const result = resolveRunDir({
			cwd: base,
			env: {},
			now: new Date("2026-04-11T10:32:15.000Z"),
		});
		assert.equal(result.source, "auto");
		assert.equal(result.label, "run");
		assert.match(result.runDir, /\/20260411T103215000Z-run$/);
	});
});

File: internal/runtime/active_run_test.go
package runtime

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

func TestResolveRunDirPrecedence(t *testing.T) {
	now := time.Date(2026, 7, 11, 10, 32, 15, 0, time.UTC)
	t.Run("explicit output wins", func(t *testing.T) {
		cwd := t.TempDir()
		outputDir := "explicit"
		result, err := ResolveRunDir(&outputDir, nil, nil, map[string]string{ActiveRunEnvVar: filepath.Join(cwd, "ignored")}, now, cwd)
		if err != nil {
			t.Fatalf("ResolveRunDir returned error: %v", err)
		}
		want := filepath.Join(cwd, outputDir)
		if result.Source != "explicit" || result.RunDir != want || result.Created {
			t.Fatalf("unexpected explicit result: %#v", result)
		}
		assertRuntimeTestDir(t, want)
	})

	t.Run("active env wins without explicit output", func(t *testing.T) {
		cwd := t.TempDir()
		active := filepath.Join(cwd, "active")
		if err := os.Mkdir(active, 0o755); err != nil {
			t.Fatalf("Mkdir returned error: %v", err)
		}
		result, err := ResolveRunDir(nil, nil, nil, map[string]string{ActiveRunEnvVar: active}, now, cwd)
		if err != nil {
			t.Fatalf("ResolveRunDir returned error: %v", err)
		}
		if result.Source != "active" || result.RunDir != active || result.Created {
			t.Fatalf("unexpected active result: %#v", result)
		}
	})

	t.Run("whitespace explicit output falls through to active env", func(t *testing.T) {
		cwd := t.TempDir()
		active := filepath.Join(cwd, "active")
		if err := os.Mkdir(active, 0o755); err != nil {
			t.Fatalf("Mkdir returned error: %v", err)
		}
		outputDir := " \t\n"
		result, err := ResolveRunDir(&outputDir, nil, nil, map[string]string{ActiveRunEnvVar: active}, now, cwd)
		if err != nil {
			t.Fatalf("ResolveRunDir returned error: %v", err)
		}
		if result.Source != "active" || result.RunDir != active || result.Created {
			t.Fatalf("unexpected active fallback result: %#v", result)
		}
	})

	for _, test := range []struct {
		name      string
		outputDir *string
		env       map[string]string
	}{
		{name: "whitespace explicit output is absent", outputDir: stringPointer(" \t\n"), env: map[string]string{}},
		{name: "whitespace active env is absent", env: map[string]string{ActiveRunEnvVar: " \t\n"}},
	} {
		t.Run(test.name, func(t *testing.T) {
			cwd := t.TempDir()
			label := "fallback"
			result, err := ResolveRunDir(test.outputDir, nil, &label, test.env, now, cwd)
			if err != nil {
				t.Fatalf("ResolveRunDir returned error: %v", err)
			}
			if result.Source != "auto" || !result.Created || result.Label == nil || *result.Label != label {
				t.Fatalf("unexpected auto result: %#v", result)
			}
			assertRuntimeTestDir(t, result.RunDir)
			if _, err := os.Stat(filepath.Join(result.RunDir, RunManifestName)); err != nil {
				t.Fatalf("expected run manifest: %v", err)
			}
		})
	}
}

func TestReadActiveRunDirRejectsInvalidTargets(t *testing.T) {
	if result, err := ReadActiveRunDir(map[string]string{ActiveRunEnvVar: " \t\n"}); err != nil || result != nil {
		t.Fatalf("whitespace active run = %v, %v; want nil, nil", result, err)
	}

	missing := filepath.Join(t.TempDir(), "missing")
	_, err := ReadActiveRunDir(map[string]string{ActiveRunEnvVar: missing})
	if err == nil || !strings.Contains(err.Error(), ActiveRunEnvVar+" does not exist") {
		t.Fatalf("expected missing active-run error, got %v", err)
	}

	filePath := filepath.Join(t.TempDir(), "file")
	if err := os.WriteFile(filePath, []byte("not a directory"), 0o644); err != nil {
		t.Fatalf("WriteFile returned error: %v", err)
	}
	_, err = ReadActiveRunDir(map[string]string{ActiveRunEnvVar: filePath})
	if err == nil || !strings.Contains(err.Error(), ActiveRunEnvVar+" must be a directory") {
		t.Fatalf("expected non-directory active-run error, got %v", err)
	}
}

func stringPointer(value string) *string {
	return &value
}

func assertRuntimeTestDir(t *testing.T, path string) {
	t.Helper()
	info, err := os.Stat(path)
	if err != nil {
		t.Fatalf("expected directory %s: %v", path, err)
	}
	if !info.IsDir() {
		t.Fatalf("expected directory %s", path)
	}
}
